### PR TITLE
Added code to allow a thread to leave the thread pool.

### PR DESCRIPTION
**Pull request by jgates108**

### CODE REVIEW

File: core/modules/util/EventThread.cc [hunk 1]
```diff
@@ -46,40 +46,98 @@ namespace util {
...
+        _currentCommand = nullptr;
     }
     finishup();
 }
 
+
+/// Ensure that commandFinish is only called once.
+void EventThread::callCommandFinish(Command::Ptr const& cmd) {
+    if (_commandFinishCalled.exchange(true) == false) {
+        _q->commandFinish(cmd);
```
Comment (andy-slac) [inline core/modules/util/EventThread.cc]: Isn't `std::call_once` doing the same with ~less typing (I never used it myself though)

File: core/modules/util/EventThread.h [hunk 2]
```diff
@@ -113,25 +114,37 @@ class EventThread : public CmdData {
...
         std::shared_ptr<MsgEnd> cmd = std::make_shared<MsgEnd>();
         queCmd(cmd);
     }
 
+    Command* getCurrentCommand() const { return _currentCommand; }
+
 protected:
     virtual void startup() {};  ///< Things to do before handling commands
     virtual void finishup() {}; ///< things to do when the thread is closing down.
+    virtual void specialActions(Command::Ptr const& cmd) {}; ///< Things to do before running a command.
```
Comment (andy-slac) [inline core/modules/util/EventThread.h]: extraneous semicolon

File: core/modules/util/EventThread.h [hunk 3]
```diff
@@ -141,11 +154,32 @@ class PoolEventThread : public EventThread {
...
 };
 
+
+/// A Command that is aware that it is running as part of a PoolEventThread,
+// which allows it to tell the pool to take special actions.
+class CommandThreadPool : public CommandTracked {
+public:
+    using Ptr = std::shared_ptr<CommandThreadPool>;
+
+    CommandThreadPool() {};
```
Comment (andy-slac) [inline core/modules/util/EventThread.h]: extraneous semicolon

File: core/modules/util/EventThread.cc (hunk 1 above)
Comment (jgates108) [inline core/modules/util/EventThread.cc]: I'll try it out and see how it goes.

Comment (jgates108) [inline core/modules/util/EventThread.cc]: Using std::call_once uses about the same amount of code. The interesting thing is that std::call_once only considers the function called if it does not throw an exception. If the function throws an exception, std::call_once will allow it be called until it does not throw an exception, and that's not what I'm looking for here.

**Comment by abh3** [comment]
This was actually pushed to master on May 11th and then orphaned.

**Pull request closed without merging**